code/modules/research/nanites/nanite_programs/sensor.dm

```diff
@@ -394,3 +394,87 @@
 	else
 		if(hearing_args[HEARING_RAW_MESSAGE] == sentence)
 			send_code()
+
+/datum/nanite_program/sensor/race
+	name = "Race Sensor"
+	desc = "When triggered, the nanites scan the host to determine their race and output a signal depending on the conditions set in the settings."
+	can_trigger = TRUE
+	trigger_cost = 0
+	trigger_cooldown = 5
+
+	extra_settings = list("Sent Code","Race","Mode")
+	var/spent = FALSE
+	var/race_type = "Human"
+	var/mode = "Is"
+	var/list/allowed_species = list(
+    	"Human" = /datum/species/human,
+    	"Lizard" = /datum/species/lizard,
+		"Moth" = /datum/species/moth,
+		"Ethereal" = /datum/species/ethereal,
+		"Pod" = /datum/species/pod,
+		"Fly" = /datum/species/fly,
+		"Felinid" = /datum/species/human/felinid,
+		"Jelly" = /datum/species/jelly
+	)
+
+/datum/nanite_program/sensor/race/set_extra_setting(user, setting)
+	if(setting == "Sent Code")
+		var/new_code = input(user, "Set the sent code (1-9999):", name, null) as null|num
+		if(isnull(new_code))
+			return
+		sent_code = CLAMP(round(new_code, 1), 1, 9999)
+	if(setting == "Race")
+		var/list/race_types = list()
+		for(var/name in allowed_species)
```

Comment:
This could be done more dynamically by doing  var/i in subtypesof(/datum/species)
then setting that to a specific species, e.g. var/datum/species/species
then getting the name by going initial(species.name)

similar things can be done throughout this code to prevent the need from that static assoc list.

Reply:
Resolved by making the allowed_species list static.